### ISSUE TITLE
Remove SQLALCHEMY_COMMIT_ON_TEARDOWN

### ISSUE THREAD

**Issue by immunda** [pallets/flask-sqlalchemy#216]

**Comment by mattupstate** [comment]
I'd guess that the reason is due to the `teardown_appcontext` callback carrying a bug that, even if you catch an exception during the app context, the `response_or_exc` will never be `None`. In other words, `teardown_appcontext` suffers from a general Python exception handling bug.

**Comment by fuhrysteve** [comment]
@mattupstate Makes sense.. it feels weird to have to commit explicitly at the end of every request.. although after doing it manually everywhere, I admit that I tend to like the fine grained control over when things get committed. I never liked the [django autocommit philosophy](https://docs.djangoproject.com/en/1.7/topics/db/transactions/#why-django-uses-autocommit).. feels like you're just trading inconvenience for sloppiness.

**Comment by immunda** [comment]
Sorry for the silence on this. Yep, that's the motivation, moving away from the (flawed) magic. I'm waiting to deprecate it entirely (3.0), because there's plans to introduce a more explicit `transaction` decorator first.

**Comment by jeffwidman** [comment]
For those who were still confused after reading the comments on this issue (like me), here's a more extensive explanation:
http://stackoverflow.com/questions/23301968/invalid-transaction-persisting-across-requests

**Comment by roganov** [comment]
Reading @jeffwidman comment helped to understand the issue, but I'm still confused what this means for flask-sqlalchemy. Is there going to be another (similar) solution or do users have to commit manually now (which would be quite unfortunate, don't want to mix persistence and application logic even more)?

**Comment by justanr** [comment]
@roganov A  handler with a try-finally would solve the problem. I'd suggest registering it as a request teardown handler and disable COMMIT_ON_TEARDOWN 

However, I'm also of the opinion that transaction management is part of persistence handling. And that persistence and application logic are two different concerns, but this isn't the forum for me going on about that.

**Comment by ben4ever** [comment]
Recently PR [pallets/flask#1822](https://github.com/pallets/flask/pull/1822) got merged into Flask.
Will this maybe change the fact whether SQLALCHEMY_COMMIT_ON_TEARDOWN will still be removed in future?

**Comment by davidism** [comment]
Yeah, it's safer now, but I really just don't like the pattern.  Explicit is better than implicit, etc.

Commit on teardown only moves the last call to `db.commit()` out of the view function, which doesn't seem like much of a burden to begin with.  It also means that the developer can no longer handle commit errors themselves, they would have to do the commit themselves in order to wrap it, at which point commit on teardown is useless.

I'm still for removing this, but open to discussion.

**Comment by asteinlein** [comment]
I disagree with removing it. Having `db.commit()` at the end of basically every view is very anti-DRY. Yes, I could add my own teardown function or some other hook, but I agree with i.e. @justanr that this belongs to persistence handling that this library is supposed to help us with.

**Comment by davidism** [comment]
That's not what DRY means.

**Comment by asteinlein** [comment]
Don't Repeat Yourself. `db.commit()` in every view is repeating yourself, when it might as well be abstracted away.

**Comment by davidism** [comment]
Committing during teardown moves an important part of the view to *after* the response is generated.  Teardown functions can't modify the response, so there's no way to indicate that the operation actually failed.

**Comment by asteinlein** [comment]
You have error handlers for that.

**Comment by davidism** [comment]
No, they run before the teardown functions.

**Comment by gdxn96** [comment]
I agree about the dislike of the pattern. However I'd much prefer to have a failsafe against lazy programmers, and have missing data associated with an error in the error log, rather than the data just be missing, requiring investigation. ~My two cents

Doing commit on teardown manually gave me the following issue
https://github.com/pallets/flask/issues/984

Basically the manual way I did commit on teardown polluted my log file with 
```
Unexpected exception [Errno 35] Resource temporarily unavailable
```
all over the place using gunicorn. Using this flag solves this issue for me, as obviously something fancy I haven't thought of is being implemented instead of my own teardown.

This was my teardown if interested
```
@app.teardown_request       
def teardown_request(exception):      
    if exception:     
        app.logger.error(exception)       
    try:      
        db.session.commit()       
    except:       
        db.session.rollback()     
    finally:      
        db.session.close()        
        db.session.remove()     
```

Sorry if this comment has no place on this thread, new enough to github 👍

**Comment by justanr** [comment]
@asteinlein I'd like to revisit the second part of my comment. Some of my opinions have changed and I also think it was misunderstood. 

I understand the desire to not do `db.commit()` in every view. But that's already the wrong place, in my opinion, to handle it. The view should should serve as just a translation device between HTTP (or rather WSGI)  and your application. For a small application, it's whatever but as an app grows,  it'll begin buckling under pressure from this design.

I prefer to silo database access into repositories that handle all persistence concerns for me. This accomplishes two things:

1. When I test, I can provide an in memory version of the repository so I don't need to mock out the SQLA queries (which is terrible to do). 

2. I have a place to put common queries. You can argue that on the model is the place to put these but I've always found this awkward. 

Fixing the proliferation of finder methods is still there but if you're interested, we can take this off channel (@just_anr on twitter). 

I said all that to say this: persistence is tricky and COMMIT_ON_TEARDOWN is unreliable at best. You should be handling commits, rollbacks and failures in your code. FSQLA probably shouldn't be making those choices for us. COMMIT_ON_TEARDOWN was a good idea, but how you handle a failed insert will change based on the data being inserted and FSQLA can't automate that for you. At least not easily.

**Comment by fuhrysteve** [comment]
My opinions have also changed since I was interested in this topic.

I agree with @justanr. `COMMIT_ON_TEARDOWN` encourages bad design that is difficult to test. Rather than explain in detail why I think this, I'll show an example and counter example.


Bad design:

```python
from .models import Widget

@app.route('/widgets/<int:widget_id>')
def update_widgets(widget_id):
    widget = db.session.query(Widget).get_or_404(widget_id)
    widget.description = request.form['description']
    db.session.commit()
    widget_json = serialize_this_somehow(widget)
    return jsonify(data=widget_json)
```


Better design:

```python
from .models import Widget

@app.route('/widgets/<int:widget_id>')
def update_widgets(widget_id):
    widget_json = widget_updater(widget_id, description=request.form['description'])
    return jsonify(data=widget_json)

def widget_updater(widget_id, description, session=None):
    session = session or db.session
    widget = db.session.query(Widget).get(widget_id)
    widget.description = description
    session.commit()
    return serialize_this_somehow(widget)
```

Now with very little adjustment, you can write a unit test for `widget_updater` that executes in 1 or 2ms instead of a 500ms integration test that requires a flask client and a database to run.

**Comment by agronholm** [comment]
This is not an easy problem to solve in any elegant fashion. Basically when the request ends, any typical web app will want to do the following:

* commit the transaction
* render a template or serialize data
* return a response

If this is done in that very order, the session by default expires all objects, forcing a reload from the database when templating or serialization is done. If the order of the two actions is reversed, it will require an explicit db.session.flush() or else you may end up serializing transient objects that don't even have their defaults applied or the primary keys are missing.

How have others solved this problem?

**Comment by ryanli1994** [comment]
It's more than two years. Anyone can tell the status quo of removing it?

**Comment by iROCKBUNNY** [comment]
3 years! Cannot believe myself that I have just finished reading all these...

**Comment by mliker** [comment]
Hi,

while I like the approach for better desing @fuhrysteve has outlined, it seems to me that this does not help with the scenario where the response is not processed by the client hence I have commited something that in the end did not happen.

Only after the client has accepted the answer would I like to commit a transaction involved.

How to does one approach that?

**Comment by agronholm** [comment]
Does any existing web framework provide such guarantees?

**Comment by mliker** [comment]
Not that I am aware of, but I am no web developer hence my scope is limited to flask where I am building a simple REST service. However my background is in telecom and finance where you simply can not afford things like that.

One thing that I am thinking of right now is whether there is an ability to catch errors during sending of a response?

**Comment by agronholm** [comment]
There is no way to guarantee that the client has received a response, since according to the HTTP protocol the client does not send any acknowledgement for the response it received. The only clue would be when the client sends the TCP level ack packets for the response but I am not aware of any way to detect that on the application (framework) level.

**Comment by agronholm** [comment]
Besides, usually the framework sits behind a reverse proxy, so relying on TCP level ACKs won't work anyway.

**Comment by agronholm** [comment]
@mliker Are you suggesting that commit be done only after the response has been sent?

**Comment by mliker** [comment]
@agronholm well, I was imagining it could be wrapped in a custom context manager where upon a successful response the context manager would do commit. However I guess the rerun just ends the execution of a function.

However I am starting to feel that I might be over-engineering this.

**Comment by agronholm** [comment]
@gdxn96 I looked at your teardown function and it has a few problems:

1. I really don't think explicitly logging the exception should be done here, but left to other parts of the framework
1. The .remove() method also closes the session, making the .close() call redundant
1. Closing the session automatically rolls back the transaction (if needed for the engine), so I would consider the .rollback() redundant too

**Comment by agronholm** [comment]
Oh, and I forgot the most important part: your teardown function tries to commit even if the application raised an exception. Do you really want that?

**Comment by davidism** [comment]
Unfortunately missed the window to deprecate this in 2.4, so this will be deprecated in 3.0 and removed in 3.1.

**Comment by ekoka** [comment]
I will go against the grain here and present some arguments in favor of committing at the end of the request. Note that I'm not advocating specifically for the reinstatement of the buggy implementation in Flask-SQLAlchemy. 

Many in the thread have taken advantage of this bug to manifest their discontent with the pattern, but the arguments presented so far seem anemic or unrelated. To those who paid attention they boil down to "I don't like it" or "it feels weird". It's unfortunate for people still undecided and pondering a way to make it work, as the silent upvotes makes it seem as though there's validity in the claim that the practice is unsound. 

But where is the evidence? 

Some of the reasons provided here rather demonstrate that there might simply be some lack of understanding (and discomfort) about the mechanics of "[unit-of-work](https://martinfowler.com/eaaCatalog/unitOfWork.html)", the pattern around which the SQLAlchemy session is [modeled](https://docs.sqlalchemy.org/en/13/glossary.html#term-unit-of-work). [Throughout](https://docs.sqlalchemy.org/en/13/faq/performance.html#i-m-inserting-400-000-rows-with-the-orm-and-it-s-really-slow) its [documentation](https://docs.sqlalchemy.org/en/13/orm/session_basics.html#what-does-the-session-do) we are [reminded](https://docs.sqlalchemy.org/en/13/orm/session_basics.html#when-do-i-construct-a-session-when-do-i-commit-it-and-when-do-i-close-it) how best to use it.

There's even this excerpt that suggests that "commit on teardown" is not only a legitimate option, but it's almost the expected practice with a "unit-of-work" framework.

> 
> It’s usually not very hard to determine the best points at which to begin and end the scope of a Session, though the wide variety of application architectures possible can introduce challenging situations.
> 
> [...]
> 
> While there’s no one-size-fits-all recommendation for how transaction scope should be determined, there are common patterns. Especially if one is writing a web application, the choice is pretty much established.
> 
> A web application is the easiest case because such an application is already constructed around a single, consistent scope - this is the request[...] Integrating web applications with the Session is then the straightforward task of linking the scope of the Session to that of the request. The Session can be established as the request begins [...] As the request ends, the Session is torn down as well, usually through the usage of event hooks provided by the web framework. [...]

It may be the case that I'm simply unaware of interesting ways that many people are using Flask. I simply use it as a wsgi framework to handle http requests. In the vast majority of cases I want _all_ database operations spanning a request to be part of the same unit-of-work. If one operation fails, it's likely that all should fail. The last thing I want in that context is some helper function "going rogue" with its own commits, while the rest is still validating the transaction. If I want something to commit on its own, then it's probably part of a different unit-of-work and should operate on its own separate transaction. Again, that last scenario would be in the slim minority.

So how does that work? I have a single transaction (one session), I flush multiple times throughout the request, I might rollback occasionally and proceed with the rest of the transaction, or I may decide to abort completely. But I generally commit only once, at the end of the request.
